diff --git a/dlls/ole32/compobj.c b/dlls/ole32/compobj.c
index d136ee623a..7a0dde3834 100644
--- a/dlls/ole32/compobj.c
+++ b/dlls/ole32/compobj.c
@@ -4547,6 +4547,7 @@ HRESULT WINAPI CoWaitForMultipleHandles(DWORD dwFlags, DWORD dwTimeout,
                     }
                 }
 
+                PeekMessageW(NULL, apt->win, 0, 0, PM_NOREMOVE|PM_NOYIELD);
                 /* some apps (e.g. Visio 2010) don't handle WM_PAINT properly and loop forever,
                  * so after processing 100 messages we go back to checking the wait handles */
                 while (count++ < 100 && COM_PeekMessage(apt, &msg))
