diff -ur a/dlls/wmp/player.c b/dlls/wmp/player.c
--- a/dlls/wmp/player.c	2019-05-24 22:45:04.000000000 +0200
+++ b/dlls/wmp/player.c	2020-03-29 23:17:15.771042878 +0200
@@ -1454,6 +1454,7 @@
     WindowsMediaPlayer *This = impl_from_IWMPControls(iface);
 
     TRACE("(%p)\n", This);
+    return NS_S_WMPCORE_COMMAND_NOT_AVAILABLE;
 
     if (!This->media) {
         return NS_S_WMPCORE_COMMAND_NOT_AVAILABLE;
