diff -ur a/dlls/winmm/mmio.c b/dlls/winmm/mmio.c
--- a/dlls/winmm/mmio.c	2019-05-24 22:45:04.000000000 +0200
+++ b/dlls/winmm/mmio.c	2020-03-28 14:40:52.085312528 +0100
@@ -76,8 +76,7 @@
 
 	    /* if filename NULL, assume open file handle in adwInfo[0] */
 	    if (szFileName) {
-                OFSTRUCT    ofs;
-                lpmmioinfo->adwInfo[0] = OpenFile(szFileName, &ofs, lpmmioinfo->dwFlags & 0xFFFF);
+                lpmmioinfo->adwInfo[0] = _lopen(szFileName, lpmmioinfo->dwFlags);
             }
 	    if (lpmmioinfo->adwInfo[0] == HFILE_ERROR)
 		ret = MMIOERR_FILENOTFOUND;
