From: Haoyang Chen <chenhaoyang@uniontech.com>
Subject: [PATCH] user32: Send WM_NCCALCSIZE message, when a new window is displayed.
Message-Id: <ffb60922-6174-902a-8de3-abaa117a3b81@uniontech.com>
Date: Tue, 12 Jan 2021 10:14:33 +0800

From 0ef95fe7663a0f35352d6ac1d66698de021b883e Mon Sep 17 00:00:00 2001
From: Haoyang Chen <chenhaoyang@uniontech.com>
Date: Mon, 11 Jan 2021 17:54:36 +0800
Subject: [PATCH] user32: Send WM_NCCALCSIZE message, when a new window is
 displayed.

Signed-off-by: Haoyang Chen <chenhaoyang@uniontech.com>
---
 dlls/user32/winpos.c | 3 +++
 1 file changed, 3 insertions(+)

diff --git a/dlls/user32/winpos.c b/dlls/user32/winpos.c
index 9e5a0c238af..51c0b822df5 100644
--- a/dlls/user32/winpos.c
+++ b/dlls/user32/winpos.c
@@ -1111,6 +1111,9 @@ static BOOL show_window( HWND hwnd, INT cmd )
             goto done;
     }
 
+    if (showFlag && !wasVisible && ((style & (WS_CAPTION | WS_MAXIMIZE | WS_MAXIMIZE)) == WS_CAPTION))
+        swp |= SWP_FRAMECHANGED;
+
     if ((showFlag != wasVisible || cmd == SW_SHOWNA) && cmd != SW_SHOWMAXIMIZED && !(swp & SWP_STATECHANGED))
     {
         SendMessageW( hwnd, WM_SHOWWINDOW, showFlag, 0 );

-- 
2.20.1

